### ISSUE TITLE
JSON support of "null" values

### ISSUE THREAD

**Issue by steweg** [CESNET/libyang#2214]
Hi, I just run into an issue that if JSON includes valid attribute name but has value set to `null`, libyang gives me various error depending of schema type. My personal expectation was that `null` values will be by default silently ignored and removed from the data model, possibly replaced with implicit default values. Therefore my question is whether this is intentional or it is a bug?

Example yang file:
```
module a {
  namespace "a";
  prefix a;
  leaf a {type string;}
}
```
JSON file:
```
{"a:a": null}
```
Output of yanglint:
```
libyang err : Invalid non-string-encoded string value "". (/a:a)
YANGLINT[E]: Failed to parse input data file "a.json".
```

**Comment by michalvasko** [comment]
> My personal expectation was that null values will be by default silently ignored and removed from the data model, possibly replaced with implicit default values.

Can you back up your expectation by something, some document ideally? I think it works this way because of XML where there always must be some value, `""` considered as a valid "no-value". I do not mind changing it for JSON but I have no idea how it should work.

**Comment by steweg** [comment]
> > My personal expectation was that null values will be by default silently ignored and removed from the data model, possibly replaced with implicit default values.
> 
> Can you back up your expectation by something, some document ideally? I think it works this way because of XML where there always must be some value, `""` considered as a valid "no-value". I do not mind changing it for JSON but I have no idea how it should work.

Unfortunately neither do I, that's why I am asking. I just know that `null` is valid JSON data type. I have seen that being allowed to be used in different implementations, which were not using `libyang` and was based on HTTP/JSON REST API, where the payload are based on YANG schema. It was used to modify existing data on server in a way, that given attribute was completely removed from the set. Even in my own implementation which uses `libyang` I have run into the similar issue due to python framework nature, which automatically translate python `None` into JSON `null` and vice versa. So I either need to filter it out manually before processing in `libyang` or allow it inside `libyang`. My preference would be to allow it within `libyang` if it doesn't violate any existing standards/rules...

**Comment by jktjkt** [comment]
Since the JSON serialization of YANG data is [well described in RFC 7951](https://datatracker.ietf.org/doc/html/rfc7951), I think there's very little room for introducing non-compatible quirks. This part of the RFC appears to be relevant:

> [6.9](https://datatracker.ietf.org/doc/html/rfc7951#section-6.9).  The "empty" Type
>
>   An "empty" value is represented as "[null]", i.e., an array with the "null" literal being its only element.  For the purposes of this document, "[null]" is considered an atomic scalar value.
>
>   This encoding of the "empty" type was chosen instead of using simply "null" in order to facilitate the use of empty leafs in common programming languages where the "null" value of a member is treated as if the member is not present.

In my opinion, the second paragraph in particular can be interpreted as "there are some languages which treat JSON this way, so let's not use that in YANG". I do not think that this means that a compliant YANG implementation should start ignoring invalid data by default.

**Comment by michalvasko** [comment]
Right, I believe this sentence in particular ([ref](https://datatracker.ietf.org/doc/html/rfc7952#section-5.2.1))

> The value of a metadata annotation SHALL be encoded in exactly the
   same way as the value of a YANG leaf node having the same type as the
   annotation;

is clear and it should already work this way.

**Comment by steweg** [comment]
> Right, I believe this sentence in particular ([ref](https://datatracker.ietf.org/doc/html/rfc7952#section-5.2.1))
> 
> > The value of a metadata annotation SHALL be encoded in exactly the
> > same way as the value of a YANG leaf node having the same type as the
> > annotation;
> 
> is clear and it should already work this way.
Well, I am not referring to `metadata`, but to data instances itself. So not sure if it is relevant or not....

> In my opinion, the second paragraph in particular can be interpreted as "there are some languages which treat JSON this way, so let's not use that in YANG". I do not think that this means that a compliant YANG implementation should start ignoring invalid data by default.
Would it ok if I would prepare a patch with new CTX option that will allow it be used directly with `null` value? So libyang will work as it is right now by default and if user would like to use this switch, he/she can do it. I believe this can be beneficial for any user which is dealing with JSON based apps, where this might be needed, so users don't need to make "translation" of one JSON to another JSON.

**Comment by michalvasko** [comment]
> Well, I am not referring to metadata, but to data instances itself. So not sure if it is relevant or not....

Oh, sorry, not sure how I got the idea we are talking about `metadata`.

> Would it ok if I would prepare a patch with new CTX option that will allow it be used directly with null value?

Well, I suppose there is no harm if it is an optional flag. But I see no reason for it to be in the context, a new (JSON) parsing flag could be added.

**Comment by steweg** [comment]
Ok, thanks for the hint of FLAG, closing this discussion, I will prepare a PR and continue discussion there.
